docs/notebooks/openai_gym_lunar_lander.pct.py

```diff
@@ -0,0 +1,253 @@
+
+# %%
+# copied verbatim from https://github.com/uber-research/TuRBO
+def heuristic_Controller(s, w):
+    angle_targ = s[0] * w[0] + s[2] * w[1]
+    if angle_targ > w[2]:
+        angle_targ = w[2]
+    if angle_targ < -w[2]:
+        angle_targ = -w[2]
+    hover_targ = w[3] * np.abs(s[0])
+
+    angle_todo = (angle_targ - s[4]) * w[4] - (s[5]) * w[5]
+    hover_todo = (hover_targ - s[1]) * w[6] - (s[3]) * w[7]
+
+    if s[6] or s[7]:
+        angle_todo = w[8]
+        hover_todo = -(s[3]) * w[9]
+
+    a = 0
+    if hover_todo > np.abs(angle_todo) and hover_todo > w[10]:
+        a = 2
+    elif angle_todo < -w[11]:
+        a = 3
+    elif angle_todo > +w[11]:
+        a = 1
+    return a
+
+
+
+STEPS_LIMIT = 1000
+TIMEOUT_REWARD = -100
+
+
+def demo_heuristic_lander(env, w, print_reward=False):
+    total_reward = 0
+    steps = 0
+    s = env.reset()
+
+    while True:
+        if steps > STEPS_LIMIT:
+            total_reward -= TIMEOUT_REWARD
+            break
+
+        a = heuristic_Controller(s, w)
+        s, r, done, info = env.step(a)
+        total_reward += r
+
+        steps += 1
+        if done:
+            break
+
+    if print_reward:
+        print(f"Total reward: {total_reward}")
+
+    return total_reward
+
+
+# %% [markdown]
+# Original values for the various parameters take values between 0.05 and 1.0. We will set the search range for each parameter to be the same, from 0.0 to 1.2.
+
+# %%
+search_space = trieste.space.Box([0.0] * 12, [1.2] * 12)
+
+# %% [markdown]
+# Let's see what kind of reward we might get by just using random parameters from this search space. Usual reward values for the Lunar Lander environment are between -250 (terrible crash) to 250 (excellent landing).
+
+# %%
+for _ in range(10):
+    sample_w = search_space.sample(1).numpy()[0]
+    demo_heuristic_lander(env, sample_w, print_reward=True)
+
+# %% [markdown]
+# As you can see, most of the random sets of parameters result in a negative reward. So picking a value from this search space at random can result in a various unwanted behaviors. Here we show some examples of the landing not going according to plan. Each of these examples was created with a sample of the parameter values from the search space.
+#
+# **Warning:** all the videos in this notebook were pre-generated. Creating renders of OpenAI Gym environments requires various dependencies depending on software setups and operating systems, so we have chosen not to do it here in the interest of transferability of this notebook. For those interested in reproducing these videos, we have saved the input parameters and the code we used to generate them in the Trieste repository, in the folder next to this notebook. However because of the stochastic nature of the environment and the optimization described here, you results might differ slightly from those shown here.
```

Comment:
```suggestion
# **Warning:** all the videos in this notebook were pre-generated. Creating renders of OpenAI Gym environments requires various dependencies depending on software setups and operating systems, so we have chosen not to do it here in the interest of transferability of this notebook. For those interested in reproducing these videos, we have saved the input parameters and the code we used to generate them in the Trieste repository, in the folder next to this notebook. However because of the stochastic nature of the environment and the optimization described here, your results might differ slightly from those shown here.
```

Reply:
someone's read this, yay!